### PR TITLE
Create-dataset is up and running

### PR DESCRIPTION
**Pull request by Antonhansel**

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -2,170 +2,100 @@
...
+  // Function called by async.auto, see below
+  var generateObject = function(model, key, cb) {
+    module.exports.config[model].generator(rawDataset[key], function(err, result) {
+      if(err) {
+        cb(err);
+      }
+      // see .defer
+      for(var subkey in result) {
+        if(result[subkey] instanceof Function) {
+          result[subkey] = result[subkey]();
```
Comment (Neamar) [inline lib/index.js]: let's send the dataset here

File: lib/index.js [hunk 2]
```diff
@@ -2,170 +2,100 @@
 
 var async = require('async');
 
+module.exports.config = {};
+module.exports.dataset = {};
```
Comment (Neamar) [inline lib/index.js]: not as a global

File: lib/index.js [hunk 3]
```diff
@@ -2,170 +2,100 @@
...
-  if(!cb) {
-    cb = waitForES;
-    waitForES = false;
+module.exports = function createDataset(rawDataset, dataset, callback) {
+  if(!callback) {
+    module.exports.dataset = {};
+    callback = dataset;
   }
+  else {
+    module.exports.dataset = dataset;
```
Comment (Neamar) [inline lib/index.js]: nope

File: lib/index.js [hunk 4]
```diff
@@ -2,170 +2,100 @@
 
 var async = require('async');
 
+module.exports.config = {};
```
Comment (Neamar) [inline lib/index.js]: you need it to set it after

File: lib/index.js [hunk 5]
```diff
@@ -2,170 +2,100 @@
...
-        else {
-          data[subkey] = objects[key][subkey];
+  // Function called by async.auto, see below
+  var generateObject = function(model, key, cb) {
+    module.exports.config[model].generator(rawDataset[key], function(err, result) {
+      if(err) {
+        cb(err);
+      }
+      // see .defer
+      for(var subkey in result) {
```
Comment (Neamar) [inline lib/index.js]: this should be done before using the generator

File: lib/index.js [hunk 6]
```diff
@@ -2,170 +2,100 @@
...
-            cb(null);
-          });
-        });
-      };
-      pusher(key, data);
-    }
-
-    async.parallel(stack, cb);
+      module.exports.dataset[key] = result;
+      cb(null);
```
Comment (Neamar) [inline lib/index.js]: cb(null, result)

File: lib/index.js [hunk 7]
```diff
@@ -2,170 +2,100 @@
...
+          if(module.exports.config[model].dependencies) {
+            asyncBuildList[key] = module.exports.config[model].dependencies;
+          }
+          // push the function for async.auto
+          asyncBuildList[key].push(function(cb) {
+            generateObject(model, key, cb);
+          });
+          cb();
+        }, cb);
+      }, cb);
```
Comment (Neamar) [inline lib/index.js]: let's carry asyncBuildList and remove it from globals

File: lib/index.js [hunk 8]
```diff
@@ -2,170 +2,100 @@
...
+        if(!matched) {
+          cb(new Error("Unable to match " + key + ". Available: " + models));
+        }
+      });
+      cb(null, buildList);
+    },
+    function buildAutoBuildList(buildList, cb) {
+      async.each(Object.keys(buildList), function(model, cb) {
+        async.each(Object.keys(buildList[model]), function(key, cb) {
+          asyncBuildList[key] = [];
```
Comment (Neamar) [inline lib/index.js]: asyncBuildList[key] = module.exports.config[model].dependencies || []

File: lib/index.js [hunk 9]
```diff
@@ -2,170 +2,100 @@
...
+          // push the dependencies of the object
+          if(module.exports.config[model].dependencies) {
+            asyncBuildList[key] = module.exports.config[model].dependencies;
```
Comment (Neamar) [inline lib/index.js]: asyncBuildList[key] = asyncBuildList[key].concat(module.exports.config[model].dependencies)

File: lib/index.js [hunk 10]
```diff
@@ -2,170 +2,100 @@
...
+        // if a model is missing
+        if(!matched) {
+          cb(new Error("Unable to match " + key + ". Available: " + models));
+        }
+      });
+      cb(null, buildList);
+    },
+    function buildAutoBuildList(buildList, cb) {
+      async.each(Object.keys(buildList), function(model, cb) {
+        async.each(Object.keys(buildList[model]), function(key, cb) {
```
Comment (Neamar) [inline lib/index.js]: Object.keys(buildList[model]).forEach

File: lib/index.js [hunk 11]
```diff
@@ -2,170 +2,100 @@
...
-      model = buildListKey;
+        // if a model is missing
+        if(!matched) {
+          cb(new Error("Unable to match " + key + ". Available: " + models));
+        }
+      });
+      cb(null, buildList);
+    },
+    function buildAutoBuildList(buildList, cb) {
+      async.each(Object.keys(buildList), function(model, cb) {
```
Comment (Neamar) [inline lib/index.js]: forEach

File: lib/index.js [hunk 12]
```diff
@@ -2,170 +2,100 @@
...
+          }
+          // push the function for async.auto
+          asyncBuildList[key].push(function(cb) {
+            generateObject(model, key, cb);
+          });
+          cb();
+        }, cb);
+      }, cb);
+    },
+    function callFinalCb() {
```
Comment (Neamar) [inline lib/index.js]: name?

File: lib/index.js [hunk 13]
```diff
@@ -2,170 +2,100 @@
...
+      // let async.auto handle the dependencies issues like a boss
+      async.auto(asyncBuildList, function(err) {
```
Comment (Neamar) [inline lib/index.js]: rarity

File: lib/index.js [hunk 14]
```diff
@@ -2,170 +2,100 @@
...
-    company: createObjectsCb('company'),
-    client: createObjectsCb('client'),
-    documentType: ['company', createObjectsCb('documentType')],
-    subcompany: ['company', createObjectsCb('subcompany', 'company')],
-    subsubcompany: ['company', 'subcompany', createObjectsCb('subsubcompany', 'company')],
-    user: ['company', 'subcompany', 'subsubcompany', createObjectsCb('user')],
-    accessToken: ['client', 'user', createObjectsCb('accessToken')],
-    document: ['documentType', 'user', 'accessToken', createObjectsCb('document')],
-  }, cb);
+    ], callback);
```
Comment (Neamar) [inline lib/index.js]: indentation

File: lib/index.js [hunk 15]
```diff
@@ -2,170 +2,100 @@
...
 };
 
-
-/**
- * Defer setting a value, waiting for another object to be created
- */
-module.exports.defer = function(dataset, key) {
+module.exports.defer = function(key) {
```
Comment (Neamar) [inline lib/index.js]: function naming

File: lib/index.js [hunk 16]
```diff
@@ -2,170 +2,100 @@
...
   if(!key) {
     throw new Error("Defer failed: missing key");
   }
-
   return function() {
```
Comment (Neamar) [inline lib/index.js]: return function(dataset)

File: test/index.js [hunk 17]
```diff
@@ -180,7 +175,7 @@ describe("createDataset(rawDataset, cb)", function() {
       }
 
       dataset.should.have.keys(['company', 'user']);
-      companyCalled.should.be.below(userCalled);
+      companyCalled.should.eql(userCalled);
```
Comment (Neamar) [inline test/index.js:179]: and usercalled should eql true

File: lib/index.js [hunk 18]
```diff
@@ -1,171 +1,100 @@
...
- *   }
- * };
- *
- * This will create a company, a user and a document and return them in dataset.
- */
-module.exports = function buildInitialState(rawDataset, dataset, waitForES, cb) {
-  if(!cb) {
-    cb = waitForES;
-    waitForES = false;
+module.exports = function createDataset(rawDataset, filledDataset, callback) {
```
Comment (Neamar) [inline lib/index.js]: function createDataset(rawDataset, dataset, callback) {
if(!callback) {
  callback = dataset;
  dataset = {}

File: lib/index.js [hunk 19]
```diff
@@ -1,171 +1,100 @@
...
-          data[subkey] = objects[key][subkey];
+  // Function called by async.auto, see below
+  var generateObject = function(model, key, cb) {
+    module.exports.config[model].generator(rawDataset[key], function(err, result) {
+      if(err) {
+        throw err;
+      }
+      // see .defer
+      for(var subkey in result) {
+        if(result[subkey] instanceof Function) {
```
Comment (Neamar) [inline lib/index.js]: still not working :(

File: lib/index.js [hunk 20]
```diff
@@ -1,171 +1,100 @@
...
-      var data = {};
-      for(var subkey in objects[key]) {
-        if(objects[key][subkey] instanceof Function) {
-          data[subkey] = objects[key][subkey]();
-        }
-        else {
-          data[subkey] = objects[key][subkey];
+  // Function called by async.auto, see below
+  var generateObject = function(model, key, cb) {
+    module.exports.config[model].generator(rawDataset[key], function(err, result) {
```
Comment (Neamar) [inline lib/index.js]: , cb)

Comment (Antonhansel) [inline lib/index.js]: Isn't it just moving the problem elsewhere? I will have to handle it line 74 anyway, right?

Comment (Neamar) [inline lib/index.js]: Yep! But i believe it's the best way to handle the problem

**Pull request merged**